source/symbols/TypeHelpers.h

```diff
@@ -0,0 +1,208 @@
+
+//------------------------------------------------------------------------------
+// TypeHelpers.h
+// Contains internal helper functions for Type implementations
+//
+// File is under the MIT license; see LICENSE for details
+//------------------------------------------------------------------------------
+#pragma once
+
+#include <numeric>
+
+#include "slang/symbols/AllTypes.h"
+#include "slang/symbols/VariableSymbols.h"
+
+namespace slang {
+
+/// Model dynamic size of a type with aR+b where a and b are compile-time constants and R is runtime
+/// determined.
+static std::pair<bitwidth_t, bitwidth_t> linearCoefficients(const Type& type,
+                                                            unsigned int destination = 0) {
+    /* IEEE standard says for dynamic-sized types:
+     // runtime check
+     struct {bit a[$]; shortint b;} a = {{1,2,3,4}, 67};
+     int b = int'(a);
+     // compile time error
+     typedef struct {byte a[$]; bit b;} dest_t;
+     int a;
+     dest_t b = dest_t'(a);
+     */
+    auto width = type.getBitWidth();
+    if (width > 0)
+        return { 0, width };
+    if (type.isString())
+        return { destination == 1 ? 0 : CHAR_BIT, 0 };
+    bitwidth_t gcd = 0;
+    // TODO: check for overflow
+    if (type.isUnpackedArray()) {
+        auto [gcd1, width1] = linearCoefficients(*type.getArrayElementType(), destination);
+        const auto& ct = type.getCanonicalType();
+        if (ct.kind == SymbolKind::FixedSizeUnpackedArrayType) {
+            auto rw = ct.as<FixedSizeUnpackedArrayType>().range.width();
+            gcd = gcd1 * rw;
+            width = width1 * rw;
+        }
+        else if (destination) {
+            if (destination > 1) {
+                if (!gcd1)
+                    gcd = width1;
+                else {
+                    gcd = gcd1;
+                    width = width1;
+                }
+            }
+        }
+        else
+            gcd = std::gcd(gcd1, width1);
+    }
+    else if (type.isUnpackedStruct()) {
+        auto& us = type.getCanonicalType().as<UnpackedStructType>();
+        for (auto& field : us.membersOfType<FieldSymbol>()) {
+            auto [gcd1, width1] = linearCoefficients(field.getType(), destination);
+            if (destination > 1 && gcd1 > 0)
+                destination--;
+            gcd = std::gcd(gcd, gcd1);
+            width += width1;
+        }
+    }
+    // TODO: classes
+    return { gcd, width };
+}
+
+static ConstantValue constContainer(const Type& type, span<const ConstantValue> elems) {
+    switch (type.kind) {
+        case SymbolKind::FixedSizeUnpackedArrayType:
+        case SymbolKind::DynamicArrayType:
+        case SymbolKind::UnpackedStructType:
+            return ConstantValue::Elements(elems.cbegin(), elems.cend());
+        case SymbolKind::QueueType:
+            return SVQueue(elems.cbegin(), elems.cend());
+        default:
+            return nullptr;
```

Comment:
Is the default case here unreachable? If so, please use the THROW_UNREACHABLE macro

Reply:
Yes, I will change it into THROW_UNREACHABLE.